# Imports

In [7]:
import pandas as pd

from sklearn.preprocessing import MultiLabelBinarizer
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split

# Dataset Inspection

In [8]:
# Load the cleaned data
df = pd.read_csv('clean_df.csv')

# Display the first few rows of the DataFrame to verify it's loaded correctly
df.head(20)

quantity  unit                        ingredient  ingredient_step  \
0       45.0    ml                          dark rum                0   
1       22.5    ml                        lime juice                1   
2       15.0    ml                             sugar                2   
3        1.0  dash                 angostura bitters                3   
4        6.0  drop                            pernod                4   
5      240.0    ml                       crushed ice                5   
6       60.0    ml                    silver tequila                0   
7       22.5    ml       marie brizard crme de cacao                1   
8       22.5    ml                       lemon juice                2   
9       45.0    ml                               gin                0   
10      22.5    ml  mari brizard white crme de cacao                1   
11      22.5    ml                      lillet blanc                2   
12      22.5    ml                       lemon juice                3   
13      45.0    ml                            pernod                0   
14       1.0   NaN                             sugar                1   
15     150.0    ml               cold filtered water                2   
16      30.0    ml                      gold tequila                0   
17      30.0    ml                          gold rum                1   
18      60.0    ml                  grapefruit juice                2   
19      90.0    ml                   pineapple juice                3   

    recipe_id  
0           0  
1           0  
2           0  
3           0  
4           0  
5           0  
6           1  
7           1  
8           1  
9           2  
10          2  
11          2  
12          2  
13          3  
14          3  
15          3  
16          4  
17          4  
18          4  
19          4

In [9]:
df.tail(5)

quantity  unit       ingredient  ingredient_step  recipe_id
2892      60.0    ml      rye whiskey                0        654
2893      22.5    ml      lemon juice                1        654
2894      22.5    ml  pineapple syrup                2        654
2895      15.0    ml      dolin blanc                3        654
2896       1.0  dash   orange bitters                4        654

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2897 entries, 0 to 2896
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   quantity         2584 non-null   float64
 1   unit             2520 non-null   object 
 2   ingredient       2896 non-null   object 
 3   ingredient_step  2897 non-null   int64  
 4   recipe_id        2897 non-null   int64  
dtypes: float64(1), int64(2), object(2)
memory usage: 113.3+ KB


In [11]:
# Explode the list of ingredients into individual rows
exploded_df = df.explode('ingredient')

# Get the list of all unique ingredients
all_ingredients = exploded_df['ingredient'].unique().tolist()

print(all_ingredients)

['dark rum', 'lime juice', 'sugar', 'angostura bitters', 'pernod', 'crushed ice', 'silver tequila', 'marie brizard crme de cacao', 'lemon juice', 'gin', 'mari brizard white crme de cacao', 'lillet blanc', 'cold filtered water', 'gold tequila', 'gold rum', 'grapefruit juice', 'pineapple juice', 'brandy', 'amontillado sherry', 'demerara syrup', 'water', 'rye whiskey', 'scotch whiskey', 'simple syrup', 'southern comfort', 'amaretto', 'sloe gin', 'orange juice', 'maraschino cherry', 'lemon wedge', 'yellow chartreuse', 'orange bitters', 'cognac', 'crme de cacao', 'cream', 'grand marnier', 'dry vermouth', 'sweet vermouth', 'bourbon whiskey', 'galliano', 'orgeat syrup', 'chilled club soda', 'white crème de cacao', 'heavy cream', 'egg white, beaten', 'light rum', 'apple schnapps', 'passion fruit liqueur', 'cranberry juice', 'apricot brandy', 'calvados', 'crème de cacao', 'crème de violette', 'chilled white crème de cacao', 'egg white', 'bénédictine', 'orange slice', 'campari', 'st. germain', '

In [13]:
# Count the frequency of each ingredient
ingredient_counts = exploded_df['ingredient'].value_counts()
# Set display options to show all rows
pd.set_option('display.max_rows', None)

# Print the full ingredient counts
print("Ingredient Counts:\n", ingredient_counts)

Ingredient Counts:
 ingredient
lime juice                                       209
lemon juice                                      178
gin                                              133
sugar                                            101
angostura bitters                                 74
simple syrup                                      71
orange juice                                      70
rye whiskey                                       61
sweet vermouth                                    57
cointreau                                         56
light rum                                         52
bourbon whiskey                                   52
vodka                                             46
dry vermouth                                      43
grenadine                                         41
pineapple juice                                   40
lemon twist                                       39
heavy cream                                       39
tequila        

# Outline:

Binary Encoding of Ingredients

In [5]:
# Assuming `recipes` is a list of lists, where each inner list contains the ingredients of a recipe
all_ingredients = list_of_all_ingredients  # List of all unique ingredients
mlb = MultiLabelBinarizer(classes=all_ingredients)
encoded_recipes = mlb.fit_transform(recipes)

# Create a DataFrame for easier manipulation
encoded_df = pd.DataFrame(encoded_recipes, columns=mlb.classes_)

NameError: name 'list_of_all_ingredients' is not defined

Model Building:

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

input_dim = len(all_ingredients)  # Number of unique ingredients

model = Sequential()
model.add(Dense(units=128, input_dim=input_dim, activation='relu'))
model.add(Dense(units=128, activation='relu'))
model.add(Dense(units=input_dim, activation='sigmoid'))  # Output layer with sigmoid activation

model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])


Model Training:

In [ ]:
X = encoded_df.values  # Input features (binary vectors of ingredients)
y = encoded_df.values  # Target labels (same as input for multi-label classification)

model.fit(X, y, epochs=50, batch_size=32, validation_split=0.2)


Generating New Recipes:

Start with a seed ingredient or set of ingredients.
Use the trained model to predict additional ingredients.

In [ ]:
seed_ingredients = ['vodka', 'orange juice']  # Example seed ingredients
seed_vector = mlb.transform([seed_ingredients])[0]

predicted_probabilities = model.predict(seed_vector.reshape(1, -1))
predicted_ingredients = mlb.inverse_transform((predicted_probabilities > 0.5).astype(int))[0]

generated_recipe = list(set(seed_ingredients + list(predicted_ingredients)))
print("Generated Recipe:", generated_recipe)


Summary of Steps:
Clean Data Preparation:

Create a DataFrame with binary encoded ingredient vectors for each recipe.
Model Building:

Define a neural network to learn from these vectors.
Model Training:

Train the model on your dataset.
Recipe Generation:

Generate new recipes starting from seed ingredients.
By following these steps, you should be able to create and train a model that can generate new cocktail recipes based on the patterns it learns from your dataset. Let me know if you need more detail on any of these steps!

The architecture described here is a simple feedforward neural network, also known as a Multi-Layer Perceptron (MLP). This architecture is used for multi-label classification, where each ingredient is treated as a binary label that can be present or absent in a recipe.

Architecture Explanation
Input Layer:

The input layer consists of nodes equal to the number of unique ingredients. Each node represents whether a particular ingredient is present (1) or absent (0) in the recipe.
Input Dimension: input_dim = len(all_ingredients).
Hidden Layers:

The network has two hidden layers, each with 128 neurons. These layers use the ReLU (Rectified Linear Unit) activation function, which introduces non-linearity to the model and allows it to learn complex patterns.
First Hidden Layer: Dense(units=128, input_dim=input_dim, activation='relu')
Second Hidden Layer: Dense(units=128, activation='relu')
Output Layer:

The output layer also consists of nodes equal to the number of unique ingredients. Each node represents the probability of the corresponding ingredient being part of the recipe.
The sigmoid activation function is used in the output layer to produce probabilities between 0 and 1 for each ingredient.
Output Layer: Dense(units=input_dim, activation='sigmoid')
Loss Function:

The model uses binary cross-entropy loss, suitable for multi-label classification where each label (ingredient) is a binary decision (present or absent).
Loss Function: binary_crossentropy
Optimizer:

The Adam optimizer is used to minimize the loss function during training. Adam is a popular choice due to its adaptive learning rate and efficiency.
Optimizer: adam
Metrics:

The accuracy metric is used to evaluate the model's performance during training and validation.
Metrics: accuracy
Why This Architecture?
Simplicity: This architecture is straightforward and easy to implement, making it suitable for a dataset of your size (654 recipes and 447 unique ingredients).
Flexibility: The use of dense layers allows the network to learn from the presence and absence of each ingredient, capturing the relationships between different ingredients.
Scalability: Adding more hidden layers or increasing the number of neurons in each layer can improve the model's capacity to learn more complex patterns, if needed.
Summary
This architecture is a feedforward neural network designed for multi-label classification. It takes a binary vector representing the presence or absence of each ingredient as input and outputs a binary vector representing the predicted probabilities of each ingredient being part of the cocktail recipe. This approach is suitable for generating new recipes by learning the common ingredient combinations from the training data.






